### PR TITLE
Lep 203 bank holiday call with faraday to avoid db table

### PR DESCRIPTION
**Pull request by ripan**
## What

[LEP-203](https://dsdmoj.atlassian.net/browse/LEP-203)

We were saving bank holidays dates in a `bank_holidays` table and refreshes table after every 10 days to get up-to-date data. As per new implementation we are making use of Faraday caching to get bank holidays dates and removing `bank_holidays` table.

[LEP-203]: https://dsdmoj.atlassian.net/browse/LEP-203?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: spec/services/govuk_bank_holiday_retriever_spec.rb [hunk 1]
```diff
@@ -2,83 +2,111 @@
...
+      stub_request(:get, "https://www.gov.uk/bank-holidays.json").to_return(body: json_response.to_json, status: 200) # .then.to_raise(StandardError)
+    end
+    let(:bank_holiday_stub_again) do
+      stub_request(:get, "https://www.gov.uk/bank-holidays.json").to_return(body: json_response.to_json, status: 200)
+    end
+
+    before do
+      bank_holiday_stub
+      Rails.cache.clear
+    end
```
Comment (starswan) [inline spec/services/govuk_bank_holiday_retriever_spec.rb:15]: I think we need to leave some notes here for future developers to explain what's going on in this test

**Comment by davidread** [comment]
>  Faraday cache expires in 1 hour.

The db caching is currently set to 10 days, and I propose we use the same figure. We will load the API enough, making calls to it every time a pod starts or restarts. And Bank holidays don't change without a lot of notice

**Comment by davidread** [comment]
> Faraday caching is working as expected but request stubbing in specs is not working well. Hence I am skipping [specs](https://github.com/ministryofjustice/cfe-civil/blob/f09984b423d99ec7d6965e5126ee0427fc6d580a/spec/services/govuk_bank_holiday_retriever_spec.rb#L23-L56) which we can fix later.

I'm just not familiar enough with this to understand what you're saying about "request stubbing in specs not working well", I'm afraid. But I believe we should expend energy on trying to test this caching, if that's a clear enough steer?

**Comment by starswan** [comment]
> > Faraday cache expires in 1 hour.
> 
> The db caching is currently set to 10 days, and I propose we use the same figure. We will load the API enough, making calls to it every time a pod starts or restarts. And Bank holidays don't change without a lot of notice

The reason for this is that the API has headers that tell us to cache for 1 hour. Overriding this back to 10 days is possible, but again it's resistant to automated testing that proves it

File: spec/services/govuk_bank_holiday_retriever_spec.rb (hunk 1 above)
Comment (ripan) [reply spec/services/govuk_bank_holiday_retriever_spec.rb]: All specs are fixed and self-explanatory, so no notes required 😄

**Comment by ripan** [comment]
> > Faraday caching is working as expected but request stubbing in specs is not working well. Hence I am skipping [specs](https://github.com/ministryofjustice/cfe-civil/blob/f09984b423d99ec7d6965e5126ee0427fc6d580a/spec/services/govuk_bank_holiday_retriever_spec.rb#L23-L56) which we can fix later.
> 
> I'm just not familiar enough with this to understand what you're saying about "request stubbing in specs not working well", I'm afraid. But I believe we should expend energy on trying to test this caching, if that's a clear enough steer?

We have updated specs to test caching. Jira comment has been removed

**Review by starswan — APPROVED** [review]

**Pull request merged**